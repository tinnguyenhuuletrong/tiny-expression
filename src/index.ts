export { parse, Parser } from "./_tinyExpressionGrammar";
export { Runtime, type RuntimeValue, type IRuntimeContext } from "./runtime";
export { deserializeFromJson, serializeToJson } from "./utils";
