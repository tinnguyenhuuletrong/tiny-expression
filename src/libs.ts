export enum ENodeType {
  Evariable,
  EConst,
  EUnaryOp,
  EUFuncOp,
  EBinaryOp,
  EArrayOp,
}
type OpType =
  | ENodeType.EUnaryOp
  | ENodeType.EUFuncOp
  | ENodeType.EBinaryOp
  | ENodeType.EArrayOp;

export enum EInstructionType {
  plus = "plus",
  minus = "minus",
  mul = "mul",
  div = "div",
  mod = "mod",
  power = "power",

  and = "and",
  or = "or",

  lessorequal = "lessorequal",
  greaterorequal = "greaterorequal",
  equal = "equal",
  notequal = "notequal",
  less = "less",
  greater = "greater",
  negate = "negate",

  contains = "contains",
  notcontains = "notcontains",
  anyof = "anyof",
  allof = "allof",
  empty = "empty",
  notempty = "notempty",
}
export type EUnaryInstructionType =
  | EInstructionType.negate
  | EInstructionType.empty
  | EInstructionType.notempty;

const ALL_UNARY_INS = [
  EInstructionType.negate,
  EInstructionType.empty,
  EInstructionType.notempty,
];

export type TargetValue = Variable | Const | Operand;

export class Variable {
  public readonly type: ENodeType = ENodeType.Evariable;
  constructor(public readonly name: string) {}

  toObject() {
    const { type, name } = this;
    return { type, name };
  }
  static fromObj(obj: any) {
    return new Variable(obj.name);
  }
}

export class Const {
  public readonly type: ENodeType = ENodeType.EConst;
  constructor(public readonly value: string | number | boolean) {}

  toObject() {
    const { type, value } = this;
    return { type, value };
  }
  static fromObj(obj: any) {
    return new Const(obj.value);
  }
}

export abstract class Operand {
  public readonly type!: OpType;
  constructor(type: OpType) {
    this.type = type;
  }
  toObject(): any {
    return { type: this.type };
  }
}

export class FunctionOperand extends Operand {
  constructor(
    public readonly name: string,
    public readonly params: ArrayOperand
  ) {
    super(ENodeType.EUFuncOp);
  }

  toObject() {
    const { name, params } = this;
    return { ...super.toObject(), name, params: params.toObject() };
  }
  static fromObj(obj: any) {
    const params = valueFactory(obj.params);
    _ensureType(params, ENodeType.EArrayOp);
    return new FunctionOperand(obj.name, params as ArrayOperand);
  }
}

export class UnaryOperand extends Operand {
  public readonly op: EUnaryInstructionType;
  public readonly target: TargetValue;

  constructor(target: TargetValue | null, op: EInstructionType) {
    super(ENodeType.EUnaryOp);

    if (!target) throw new Error(`UnaryOperand target is null`);
    if (!ALL_UNARY_INS.includes(op))
      throw new Error(`UnaryOperand not accept insType ${op}`);

    this.target = target;
    this.op = op as any;
  }

  toObject() {
    const { op, target } = this;
    return { ...super.toObject(), op, target: target.toObject() };
  }
  static fromObj(obj: any) {
    const target = valueFactory(obj.target);
    return new UnaryOperand(target, obj.op);
  }
}

export class BinaryOperand extends Operand {
  public readonly op: EInstructionType;
  public readonly left: TargetValue;
  public readonly right: TargetValue;
  public readonly isArithmeticOp: boolean;
  constructor(
    op: EInstructionType,
    left: TargetValue,
    right: TargetValue,
    isArithmeticOp: boolean = false
  ) {
    super(ENodeType.EBinaryOp);
    this.op = op;
    this.left = left;
    this.right = right;
    this.isArithmeticOp = isArithmeticOp;
  }

  toObject() {
    const { op, left, right, isArithmeticOp } = this;
    return {
      ...super.toObject(),
      op,
      left: left.toObject(),
      right: right.toObject(),
      isArithmeticOp,
    };
  }
  static fromObj(obj: any) {
    const op = obj.op;
    const left = valueFactory(obj.left);
    const right = valueFactory(obj.right);
    const isArithmeticOp = Boolean(obj.isArithmeticOp);
    return new BinaryOperand(op, left, right, isArithmeticOp);
  }
}

export class ArrayOperand extends Operand {
  public readonly data: TargetValue[];
  constructor(params: TargetValue[]) {
    super(ENodeType.EArrayOp);
    this.data = params;
  }

  toObject() {
    const { data } = this;
    return {
      ...super.toObject(),
      data: data.map((itm) => itm.toObject()),
    };
  }
  static fromObj(obj: any) {
    if (!Array.isArray(obj.data))
      throw new Error("ArrayOperand parse error. obj.data must be array");
    const data = obj.data.map((itm: any) => valueFactory(itm));
    return new ArrayOperand(data);
  }
}

export function buildBinaryOperand(
  head: any,
  tail: Array<any>,
  isArithmeticOp: boolean = false
): TargetValue {
  return tail.reduce(function (result, elements) {
    return new BinaryOperand(
      elements.v1.value,
      result,
      elements.v2.value,
      isArithmeticOp
    );
  }, head.value);
}

export function flattenArray(array: Array<any>): Array<any> {
  return [].concat.apply([], array);
}

function _ensureType(obj: any, t: ENodeType) {
  if (obj?.type !== t) throw new Error(`expected type ${t}`);
}

export function valueFactory(obj: any): TargetValue {
  const type = obj.type as ENodeType;
  switch (type) {
    case ENodeType.EConst:
      return Const.fromObj(obj);
    case ENodeType.Evariable:
      return Variable.fromObj(obj);
    case ENodeType.EUnaryOp:
      return UnaryOperand.fromObj(obj);
    case ENodeType.EArrayOp:
      return ArrayOperand.fromObj(obj);
    case ENodeType.EBinaryOp:
      return BinaryOperand.fromObj(obj);
    case ENodeType.EUFuncOp:
      return FunctionOperand.fromObj(obj);
    default:
      // should not reachable
      ((_x: never) => {})(type);
      throw new Error(`unknown type ${type}`);
  }
}
