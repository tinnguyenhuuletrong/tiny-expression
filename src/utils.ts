import type { RuntimeValue } from ".";
import { TargetValue, valueFactory } from "./libs";

export function serializeToJson(root: TargetValue): string {
  return JSON.stringify(root.toObject());
}

export function deserializeFromJson(str: string): TargetValue | null {
  const obj = JSON.parse(str);
  if (obj === null || obj === undefined) return null;
  return valueFactory(obj);
}

export function _castToStringOrNumber(v: RuntimeValue): string | number {
  const tmp = typeof v;
  if (tmp === "number" || tmp === "string") return v as string | number;
  throw new Error(`${v} is not TypePrimitive. it is ${tmp}`);
}

export function _castToNumber(v: RuntimeValue): number {
  const tmp = typeof v;
  if (tmp === "number") return v as number;
  throw new Error(`${v} is not TypePrimitive`);
}

export function _castToBoolean(v: RuntimeValue): boolean {
  return Boolean(v);
}

export function _castToDate(v: RuntimeValue): Date {
  let d!: Date;
  if (typeof v === "number") {
    d = new Date(v);
  } else if (typeof v === "string") {
    d = new Date(v);
    if (isNaN(d.valueOf())) {
      throw new Error(`${v} - ${d} is not a Date like string`);
    }
  } else {
    throw new Error(`${v} is not numer or string. can not cast to date`);
  }
  return d;
}

export function _castToArray(v: RuntimeValue): Array<RuntimeValue> {
  if (Array.isArray(v)) return v;
  throw new Error("ensureArray invalid cast");
}
