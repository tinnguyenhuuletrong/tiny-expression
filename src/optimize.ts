import {
  ArrayOperand,
  BinaryOperand,
  Const,
  EInstructionType,
  ENodeType,
  FunctionOperand,
  TargetValue,
  UnaryOperand,
} from "./libs";
import { _castToArray, _castToBoolean, _castToStringOrNumber } from "./utils";

export function doOptimize(node: TargetValue): TargetValue {
  return _doOptimizeInternal(node);
}

function _doOptimizeInternal(node: TargetValue): TargetValue {
  const type = node.type;
  switch (type) {
    case ENodeType.EBinaryOp: {
      const tmp = node as BinaryOperand;
      const left = _doOptimizeInternal(tmp.left);
      const right = _doOptimizeInternal(tmp.right);

      const canOptimize = isConst(left) && isConst(right);
      if (!canOptimize) {
        return new BinaryOperand(tmp.op, left, right, tmp.isArithmeticOp);
      }

      switch (tmp.op) {
        case EInstructionType.plus: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 + v2);
        }
        case EInstructionType.minus: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 - v2);
        }
        case EInstructionType.mul: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 * v2);
        }
        case EInstructionType.div: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 / v2);
        }
        case EInstructionType.mod: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 % v2);
        }
        case EInstructionType.power: {
          const v1 = _castToStringOrNumber((left as Const).value) as any;
          const v2 = _castToStringOrNumber((right as Const).value) as any;
          return new Const(v1 ** v2);
        }

        default:
          return node;
      }
    }
    case ENodeType.EUnaryOp: {
      const tmp = node as UnaryOperand;
      const target = _doOptimizeInternal(tmp.target);

      const canOptimize = isConst(target);
      if (!canOptimize) return node;

      switch (tmp.op) {
        case EInstructionType.negate:
          const v = _castToBoolean((target as Const).value);
          return new Const(!v);

        default:
          return node;
      }
    }
    case ENodeType.EArrayOp: {
      const ins = node as ArrayOperand;
      const newData = ins.data.map((itm) => _doOptimizeInternal(itm));
      return new ArrayOperand(newData);
    }
    case ENodeType.EUFuncOp: {
      const ins = node as FunctionOperand;
      const params = _doOptimizeInternal(ins.params);
      return new FunctionOperand(ins.name, params as ArrayOperand);
    }
    default:
      return node;
  }
}

function isConst(node: TargetValue) {
  return node.type === ENodeType.EConst;
}
