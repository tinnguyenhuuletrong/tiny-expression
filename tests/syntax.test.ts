import { test, expect } from "bun:test";
import { Parser, parse } from "../src";
import { inspect } from "util";

test("atoms cases", async () => {
  const CASES = [
    `12345`,
    `1.32`,
    `-9.99`,
    `0`,
    `100`,
    `true`,
    `false`,
    `0x1234`,
    `-0xff`,
    `-10`,
    `null`,
    `undefined`,
    `{variable_name}`,
    `{variable_array.a.b.c.d}`,
    `[1,2,3]`,
    `[1.99, {v_a}]`,
    `[0x123, is_admin(3,7)]`,
    `''`,
    `""`,
    `'i am string'`,
    `"i am also a string 12345"`,
  ];
  for (const it of CASES) {
    const res = parse(it);
    console.log(`${it} ->`, inspect(res.ast?.value, { depth: 10 }));

    if (!["null", "undefined"].includes(it))
      expect(res.ast?.value).not.toBe(null);

    expect(res.ast?.value).toMatchSnapshot(undefined);
  }
});

test("expressions cases", async () => {
  const CASES = [
    `11 + 22`,
    `0 - 31`,
    `-31 + 31`,
    `1 + 2*3`,
    `12 / (9 /0x33)`,
    `33^2`,
    `33%99`,
    `(((1 + 3) power (5 - 2)) ^ (9 / 7)) / (3 % 2)`,
    `{variable_a} ^ ({variable_b} + 1)`,
  ];

  for (const it of CASES) {
    const res = parse(it);
    console.log(`${it} ->`, inspect(res.ast?.value, { depth: 10 }));
    expect(res.ast?.value).not.toBe(null);
    expect(res.ast?.value).toMatchSnapshot();
  }
});

test("logical cases", async () => {
  const CASES = [
    `1 > 2`,
    `{variable_a} >= {variable_b}`,
    `10 / 2 < (2 * 3 + 1)`,
    `{age} - 18 < 0`,
    `{age} <= 0xff`,
    `{variable_a} == 1`,
    `{variable_b} != true`,
    `!{variable_c}`,
    `{array_a} empty`,
    `{array_b} notempty`,
    `{array_a} contain 5`,
    `{array_a} contains 5`,
    `{array_a} notcontains 5`,
    `[3,5,7] anyof 3`,
    `{array_a} allof true`,
  ];

  for (const it of CASES) {
    const res = parse(it);
    console.log(`${it} ->`, inspect(res.ast?.value, { depth: 10 }));
    expect(res.ast?.value).not.toBe(null);
    expect(res.ast?.value).toMatchSnapshot();
  }
});

test("logical cases 2", async () => {
  const CASES = [
    `{variable_a} > 0 and {variable_b}`,
    `(3 * {age} < 99) or {variable_b}`,
    `(
        {scopes} contain allow_read 
        && 
        {acc_type} == normal
      ) 
      || is_admin()`,
  ];

  for (const it of CASES) {
    const res = parse(it);
    console.log(`${it} ->`, inspect(res.ast?.value, { depth: 10 }));
    expect(res.ast?.value).not.toBe(null);
    expect(res.ast?.value).toMatchSnapshot();
  }
});

test("function cases", async () => {
  const CASES = [
    `this_function()`,
    `thatFunction({variable_a})`,
    `parseInt({variable_a}, 10)`,
    `addDate(now(), 10)`,
    `iif(
      age({day_of_birth}) > 18, 
      true, 
      false
    )`,
  ];

  for (const it of CASES) {
    const res = parse(it);
    console.log(`${it} ->`, inspect(res.ast?.value, { depth: 10 }));
    expect(res.ast?.value).not.toBe(null);
    expect(res.ast?.value).toMatchSnapshot();
  }
});

test("syntax error cases", async () => {
  const CASES = [`(0fff`, `func)`, `a <> b`];
  for (const it of CASES) {
    const res = parse(it);
    console.log(`${it} ->`, inspect(res.ast?.value, { depth: 10 }));
    expect(res.ast === null).toBe(true);
    expect(res.errs.map((itm) => itm.pos)).toMatchSnapshot();
  }
});

test("usage", async () => {
  const inp = `age( {dob} ) > 18`;
  const ins = new Parser(inp);
  const isValid = ins.test();
  if (!isValid) throw new Error("OoO");
  const res = ins.parse().ast?.value;
  console.log(inp);
  console.dir(res, { depth: 10 });
  console.log(JSON.stringify(res));
});

test("syntax debug", async () => {
  const res = parse(`100`);
  // console.dir(
  //   res.errs.map((itm) => itm.toString()),
  //   {
  //     depth: 10,
  //   }
  // );
  // console.dir(res.ast, {
  //   depth: 20,
  // });
  console.dir(res.ast?.value, {
    depth: 10,
  });
});
