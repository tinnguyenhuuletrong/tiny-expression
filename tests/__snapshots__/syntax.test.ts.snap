// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms cases 1`] = `
Const {
  "type": 1,
  "value": 12345,
}
`;

exports[`atoms cases 2`] = `
Const {
  "type": 1,
  "value": 1.32,
}
`;

exports[`atoms cases 3`] = `
Const {
  "type": 1,
  "value": -9.99,
}
`;

exports[`atoms cases 4`] = `
Const {
  "type": 1,
  "value": 0,
}
`;

exports[`atoms cases 5`] = `
Const {
  "type": 1,
  "value": 100,
}
`;

exports[`atoms cases 6`] = `
Const {
  "type": 1,
  "value": true,
}
`;

exports[`atoms cases 7`] = `
Const {
  "type": 1,
  "value": false,
}
`;

exports[`atoms cases 8`] = `
Const {
  "type": 1,
  "value": 4660,
}
`;

exports[`atoms cases 9`] = `
Const {
  "type": 1,
  "value": -255,
}
`;

exports[`atoms cases 10`] = `
Const {
  "type": 1,
  "value": -10,
}
`;

exports[`atoms cases 11`] = `null`;

exports[`atoms cases 12`] = `null`;

exports[`atoms cases 13`] = `
Variable {
  "name": "variable_name",
  "type": 0,
}
`;

exports[`atoms cases 14`] = `
Variable {
  "name": "variable_array.a.b.c.d",
  "type": 0,
}
`;

exports[`atoms cases 15`] = `
ArrayOperand {
  "data": [
    Const {
      "type": 1,
      "value": 1,
    },
    Const {
      "type": 1,
      "value": 2,
    },
    Const {
      "type": 1,
      "value": 3,
    },
  ],
  "type": 5,
}
`;

exports[`atoms cases 16`] = `
ArrayOperand {
  "data": [
    Const {
      "type": 1,
      "value": 1.99,
    },
    Variable {
      "name": "v_a",
      "type": 0,
    },
  ],
  "type": 5,
}
`;

exports[`atoms cases 17`] = `
ArrayOperand {
  "data": [
    Const {
      "type": 1,
      "value": 291,
    },
    FunctionOperand {
      "name": "is_admin",
      "params": ArrayOperand {
        "data": [
          Const {
            "type": 1,
            "value": 3,
          },
          Const {
            "type": 1,
            "value": 7,
          },
        ],
        "type": 5,
      },
      "type": 3,
    },
  ],
  "type": 5,
}
`;

exports[`atoms cases 18`] = `
Const {
  "type": 1,
  "value": "",
}
`;

exports[`atoms cases 19`] = `
Const {
  "type": 1,
  "value": "",
}
`;

exports[`atoms cases 20`] = `
Const {
  "type": 1,
  "value": "i am string",
}
`;

exports[`atoms cases 21`] = `
Const {
  "type": 1,
  "value": "i am also a string 12345",
}
`;

exports[`expressions cases 1`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 11,
  },
  "op": "plus",
  "right": Const {
    "type": 1,
    "value": 22,
  },
  "type": 4,
}
`;

exports[`expressions cases 2`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 0,
  },
  "op": "minus",
  "right": Const {
    "type": 1,
    "value": 31,
  },
  "type": 4,
}
`;

exports[`expressions cases 3`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": -31,
  },
  "op": "plus",
  "right": Const {
    "type": 1,
    "value": 31,
  },
  "type": 4,
}
`;

exports[`expressions cases 4`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 1,
  },
  "op": "plus",
  "right": BinaryOperand {
    "isArithmeticOp": true,
    "left": Const {
      "type": 1,
      "value": 2,
    },
    "op": "mul",
    "right": Const {
      "type": 1,
      "value": 3,
    },
    "type": 4,
  },
  "type": 4,
}
`;

exports[`expressions cases 5`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 12,
  },
  "op": "div",
  "right": BinaryOperand {
    "isArithmeticOp": true,
    "left": Const {
      "type": 1,
      "value": 9,
    },
    "op": "div",
    "right": Const {
      "type": 1,
      "value": 51,
    },
    "type": 4,
  },
  "type": 4,
}
`;

exports[`expressions cases 6`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 33,
  },
  "op": "power",
  "right": Const {
    "type": 1,
    "value": 2,
  },
  "type": 4,
}
`;

exports[`expressions cases 7`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Const {
    "type": 1,
    "value": 33,
  },
  "op": "mod",
  "right": Const {
    "type": 1,
    "value": 99,
  },
  "type": 4,
}
`;

exports[`expressions cases 8`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": BinaryOperand {
    "isArithmeticOp": true,
    "left": BinaryOperand {
      "isArithmeticOp": true,
      "left": BinaryOperand {
        "isArithmeticOp": true,
        "left": Const {
          "type": 1,
          "value": 1,
        },
        "op": "plus",
        "right": Const {
          "type": 1,
          "value": 3,
        },
        "type": 4,
      },
      "op": "power",
      "right": BinaryOperand {
        "isArithmeticOp": true,
        "left": Const {
          "type": 1,
          "value": 5,
        },
        "op": "minus",
        "right": Const {
          "type": 1,
          "value": 2,
        },
        "type": 4,
      },
      "type": 4,
    },
    "op": "power",
    "right": BinaryOperand {
      "isArithmeticOp": true,
      "left": Const {
        "type": 1,
        "value": 9,
      },
      "op": "div",
      "right": Const {
        "type": 1,
        "value": 7,
      },
      "type": 4,
    },
    "type": 4,
  },
  "op": "div",
  "right": BinaryOperand {
    "isArithmeticOp": true,
    "left": Const {
      "type": 1,
      "value": 3,
    },
    "op": "mod",
    "right": Const {
      "type": 1,
      "value": 2,
    },
    "type": 4,
  },
  "type": 4,
}
`;

exports[`expressions cases 9`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": Variable {
    "name": "variable_a",
    "type": 0,
  },
  "op": "power",
  "right": BinaryOperand {
    "isArithmeticOp": true,
    "left": Variable {
      "name": "variable_b",
      "type": 0,
    },
    "op": "plus",
    "right": Const {
      "type": 1,
      "value": 1,
    },
    "type": 4,
  },
  "type": 4,
}
`;

exports[`logical cases 1`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Const {
    "type": 1,
    "value": 1,
  },
  "op": "greater",
  "right": Const {
    "type": 1,
    "value": 2,
  },
  "type": 4,
}
`;

exports[`logical cases 2`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "variable_a",
    "type": 0,
  },
  "op": "greaterorequal",
  "right": Variable {
    "name": "variable_b",
    "type": 0,
  },
  "type": 4,
}
`;

exports[`logical cases 3`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": BinaryOperand {
    "isArithmeticOp": true,
    "left": Const {
      "type": 1,
      "value": 10,
    },
    "op": "div",
    "right": Const {
      "type": 1,
      "value": 2,
    },
    "type": 4,
  },
  "op": "less",
  "right": BinaryOperand {
    "isArithmeticOp": true,
    "left": BinaryOperand {
      "isArithmeticOp": true,
      "left": Const {
        "type": 1,
        "value": 2,
      },
      "op": "mul",
      "right": Const {
        "type": 1,
        "value": 3,
      },
      "type": 4,
    },
    "op": "plus",
    "right": Const {
      "type": 1,
      "value": 1,
    },
    "type": 4,
  },
  "type": 4,
}
`;

exports[`logical cases 4`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": BinaryOperand {
    "isArithmeticOp": true,
    "left": Variable {
      "name": "age",
      "type": 0,
    },
    "op": "minus",
    "right": Const {
      "type": 1,
      "value": 18,
    },
    "type": 4,
  },
  "op": "less",
  "right": Const {
    "type": 1,
    "value": 0,
  },
  "type": 4,
}
`;

exports[`logical cases 5`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "age",
    "type": 0,
  },
  "op": "lessorequal",
  "right": Const {
    "type": 1,
    "value": 255,
  },
  "type": 4,
}
`;

exports[`logical cases 6`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "variable_a",
    "type": 0,
  },
  "op": "equal",
  "right": Const {
    "type": 1,
    "value": 1,
  },
  "type": 4,
}
`;

exports[`logical cases 7`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "variable_b",
    "type": 0,
  },
  "op": "notequal",
  "right": Const {
    "type": 1,
    "value": true,
  },
  "type": 4,
}
`;

exports[`logical cases 8`] = `
UnaryOperand {
  "op": "negate",
  "target": Variable {
    "name": "variable_c",
    "type": 0,
  },
  "type": 2,
}
`;

exports[`logical cases 9`] = `
UnaryOperand {
  "op": "empty",
  "target": Variable {
    "name": "array_a",
    "type": 0,
  },
  "type": 2,
}
`;

exports[`logical cases 10`] = `
UnaryOperand {
  "op": "notempty",
  "target": Variable {
    "name": "array_b",
    "type": 0,
  },
  "type": 2,
}
`;

exports[`logical cases 11`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "array_a",
    "type": 0,
  },
  "op": "contains",
  "right": Const {
    "type": 1,
    "value": 5,
  },
  "type": 4,
}
`;

exports[`logical cases 12`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "array_a",
    "type": 0,
  },
  "op": "contains",
  "right": Const {
    "type": 1,
    "value": 5,
  },
  "type": 4,
}
`;

exports[`logical cases 13`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "array_a",
    "type": 0,
  },
  "op": "notcontains",
  "right": Const {
    "type": 1,
    "value": 5,
  },
  "type": 4,
}
`;

exports[`logical cases 14`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": ArrayOperand {
    "data": [
      Const {
        "type": 1,
        "value": 3,
      },
      Const {
        "type": 1,
        "value": 5,
      },
      Const {
        "type": 1,
        "value": 7,
      },
    ],
    "type": 5,
  },
  "op": "anyof",
  "right": Const {
    "type": 1,
    "value": 3,
  },
  "type": 4,
}
`;

exports[`logical cases 15`] = `
BinaryOperand {
  "isArithmeticOp": false,
  "left": Variable {
    "name": "array_a",
    "type": 0,
  },
  "op": "allof",
  "right": Const {
    "type": 1,
    "value": true,
  },
  "type": 4,
}
`;

exports[`logical cases 2 1`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": BinaryOperand {
    "isArithmeticOp": false,
    "left": Variable {
      "name": "variable_a",
      "type": 0,
    },
    "op": "greater",
    "right": Const {
      "type": 1,
      "value": 0,
    },
    "type": 4,
  },
  "op": "and",
  "right": Variable {
    "name": "variable_b",
    "type": 0,
  },
  "type": 4,
}
`;

exports[`logical cases 2 2`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": BinaryOperand {
    "isArithmeticOp": false,
    "left": BinaryOperand {
      "isArithmeticOp": true,
      "left": Const {
        "type": 1,
        "value": 3,
      },
      "op": "mul",
      "right": Variable {
        "name": "age",
        "type": 0,
      },
      "type": 4,
    },
    "op": "less",
    "right": Const {
      "type": 1,
      "value": 99,
    },
    "type": 4,
  },
  "op": "or",
  "right": Variable {
    "name": "variable_b",
    "type": 0,
  },
  "type": 4,
}
`;

exports[`logical cases 2 3`] = `
BinaryOperand {
  "isArithmeticOp": true,
  "left": BinaryOperand {
    "isArithmeticOp": true,
    "left": BinaryOperand {
      "isArithmeticOp": false,
      "left": Variable {
        "name": "scopes",
        "type": 0,
      },
      "op": "contains",
      "right": Const {
        "type": 1,
        "value": "allow_read",
      },
      "type": 4,
    },
    "op": "and",
    "right": BinaryOperand {
      "isArithmeticOp": false,
      "left": Variable {
        "name": "acc_type",
        "type": 0,
      },
      "op": "equal",
      "right": Const {
        "type": 1,
        "value": "normal",
      },
      "type": 4,
    },
    "type": 4,
  },
  "op": "or",
  "right": FunctionOperand {
    "name": "is_admin",
    "params": ArrayOperand {
      "data": [],
      "type": 5,
    },
    "type": 3,
  },
  "type": 4,
}
`;

exports[`function cases 1`] = `
FunctionOperand {
  "name": "this_function",
  "params": ArrayOperand {
    "data": [],
    "type": 5,
  },
  "type": 3,
}
`;

exports[`function cases 2`] = `
FunctionOperand {
  "name": "thatFunction",
  "params": ArrayOperand {
    "data": [
      Variable {
        "name": "variable_a",
        "type": 0,
      },
    ],
    "type": 5,
  },
  "type": 3,
}
`;

exports[`function cases 3`] = `
FunctionOperand {
  "name": "parseInt",
  "params": ArrayOperand {
    "data": [
      Variable {
        "name": "variable_a",
        "type": 0,
      },
      Const {
        "type": 1,
        "value": 10,
      },
    ],
    "type": 5,
  },
  "type": 3,
}
`;

exports[`function cases 4`] = `
FunctionOperand {
  "name": "addDate",
  "params": ArrayOperand {
    "data": [
      FunctionOperand {
        "name": "now",
        "params": ArrayOperand {
          "data": [],
          "type": 5,
        },
        "type": 3,
      },
      Const {
        "type": 1,
        "value": 10,
      },
    ],
    "type": 5,
  },
  "type": 3,
}
`;

exports[`function cases 5`] = `
FunctionOperand {
  "name": "iif",
  "params": ArrayOperand {
    "data": [
      BinaryOperand {
        "isArithmeticOp": false,
        "left": FunctionOperand {
          "name": "age",
          "params": ArrayOperand {
            "data": [
              Variable {
                "name": "day_of_birth",
                "type": 0,
              },
            ],
            "type": 5,
          },
          "type": 3,
        },
        "op": "greater",
        "right": Const {
          "type": 1,
          "value": 18,
        },
        "type": 4,
      },
      Const {
        "type": 1,
        "value": true,
      },
      Const {
        "type": 1,
        "value": false,
      },
    ],
    "type": 5,
  },
  "type": 3,
}
`;

exports[`syntax error cases 1`] = `
[
  {
    "line": 1,
    "offset": 5,
    "overallPos": 5,
  },
]
`;

exports[`syntax error cases 2`] = `
[
  {
    "line": 1,
    "offset": 4,
    "overallPos": 4,
  },
]
`;

exports[`syntax error cases 3`] = `
[
  {
    "line": 1,
    "offset": 3,
    "overallPos": 3,
  },
]
`;
